kotlinx-coroutines-core/common/src/channels/BroadcastChannel.kt

```diff
@@ -70,6 +74,333 @@ public fun <E> BroadcastChannel(capacity: Int): BroadcastChannel<E> =
+    // #############################
+
+    /**
+     * Sends the specified element to all subscribers.
+     *
+     * **!!! THIS IMPLEMENTATION IS NOT LINEARIZABLE !!!**
+     *
+     * As the operation should send the element to multiple
+     * subscribers simultaneously, it is non-trivial to
+     * implement it in an atomic way. Specifically, this
+     * would require a special implementation that does
+     * not transfer the element until all parties are able
+     * to resume it (this `send(..)` can be cancelled
+     * or the broadcast can become closed in the meantime).
+     * As broadcasts are obsolete, we keep this implementation
+     * as simple as possible, allowing non-linearizability
+     * in corner cases.
+     */
+    override suspend fun send(element: E) {
+        val subs = lock.withLock { // protected by lock
+            // Is this channel closed for send?
+            if (isClosedForSend) throw sendException
+            // Update the last sent element if this broadcast is conflated.
+            if (capacity == CONFLATED) lastConflatedElement = element
+            // Get a copy of subscriber list. Unfortunately,
+            // it is impossible to send the element to the subscribers
+            // under the lock due to possible suspensions.
+            ArrayList(subscribers)
+        }
+        // The lock has been released. Send the element to the
+        // subscribers one-by-one, and finish immediately
+        // when this broadcast discovered in the closed state.
+        // Note that this implementation is non-linearizable;
+        // see this method documentation for details.
+        subs.forEach {
+            // We use special function to send the element,
+            // which returns `true` on success and `false`
+            // if the subscriber is closed.
+            val success = it.sendBroadcast(element)
+            // The sending attempt has failed.
+            // Check whether the broadcast is closed.
+            if (!success && isClosedForSend) throw sendException
+        }
+    }
+
+    override fun trySend(element: E): ChannelResult<Unit> = lock.withLock { // protected by lock
+        // Is this channel closed for send?
+        if (isClosedForSend) return super.trySend(element)
+        // Check whether the plain `send(..)` operation
+        // should suspend and fail in this case.
+        val shouldSuspend = subscribers.any { it.shouldSendSuspend() }
+        if (shouldSuspend) return ChannelResult.failure()
+        // Send the element to all subscribers.
+        // It is guaranteed that the attempt cannot fail,
+        // as both the broadcast closing and subscription
+        // cancellation are guarded by lock, which is held
+        // by the current operation.
+        subscribers.forEach { it.trySend(element) }
+        // Update the last sent element if this broadcast is conflated.
+        if (capacity == CONFLATED) lastConflatedElement = element
```

Comment:
shouldn't it go before `trySend` loop? See my previous comment about reentrancy where such difference can be observed for the additional context

Reply:
It indeed would make the code slightly safer in case of reentrancy, I have applied the suggested change